File: backend/routes/scenario_prompts.py
"""
Route per gestione scenario prompts
"""
from fastapi import APIRouter, HTTPException, Depends, status
from pydantic import BaseModel
from typing import Optional, List
from uuid import UUID
from supabase import Client
from backend.database import get_supabase
from backend.middleware.auth import get_current_shop_owner
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/scenario-prompts", tags=["scenario-prompts"])


class ScenarioPromptCreate(BaseModel):
    shop_id: UUID
    name: str
    description: str  # Descrizione posizione, ambiente, illuminazione, ecc.
    position: Optional[str] = None  # Es: "in piedi", "seduto", "camminando"
    environment: Optional[str] = None  # Es: "interno", "esterno", "spiaggia", "città"
    lighting: Optional[str] = None  # Es: "naturale", "artificiale", "serale"
    background: Optional[str] = None  # Descrizione sfondo


class ScenarioPromptUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    position: Optional[str] = None
    environment: Optional[str] = None
    lighting: Optional[str] = None
    background: Optional[str] = None


@router.get("/")
async def list_scenario_prompts(
    shop_id: Optional[UUID] = None,
    current_user: dict = Depends(get_current_shop_owner),
    supabase: Client = Depends(get_supabase)
):
    """Lista scenario prompts con filtri opzionali"""
    try:
        query = supabase.table("scenario_prompts").select("*")
        
        if shop_id:
            # Verifica che il negozio appartenga al negoziante
            shop_result = supabase.table("shops").select("owner_id").eq("id", str(shop_id)).single().execute()
            if not shop_result.data or shop_result.data["owner_id"] != current_user["id"]:
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail="Accesso negato a questo negozio"
                )
            query = query.eq("shop_id", str(shop_id))
        else:
            # Se non specificato shop_id, mostra solo scenari dei negozi del negoziante
            shops_result = supabase.table("shops").select("id").eq("owner_id", current_user["id"]).execute()
            shop_ids = [shop["id"] for shop in shops_result.data]
            if shop_ids:
                query = query.in_("shop_id", [str(sid) for sid in shop_ids])
            else:
                return {"scenarios": [], "count": 0}
        
        result = query.execute()
        return {
            "scenarios": result.data,
            "count": len(result.data)
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Errore lista scenario prompts: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Errore durante il recupero scenario prompts: {str(e)}"
        )


@router.get("/{scenario_id}")
async def get_scenario_prompt(
    scenario_id: UUID,
    current_user: dict = Depends(get_current_shop_owner),
    supabase: Client = Depends(get_supabase)
):
    """Ottieni dettagli di uno scenario prompt"""
    try:
        result = supabase.table("scenario_prompts").select("*").eq("id", str(scenario_id)).execute()
        
        if not result.data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Scenario prompt non trovato"
            )
        
        scenario = result.data[0]
        
        # Verifica permessi
        shop_result = supabase.table("shops").select("owner_id").eq("id", scenario["shop_id"]).single().execute()
        if not shop_result.data or shop_result.data["owner_id"] != current_user["id"]:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Accesso negato"
            )
        
        return {"scenario": scenario}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Errore recupero scenario prompt: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Errore durante il recupero scenario prompt: {str(e)}"
        )


@router.post("/", status_code=status.HTTP_201_CREATED)
async def create_scenario_prompt(
    scenario: ScenarioPromptCreate,
    current_user: dict = Depends(get_current_shop_owner),
    supabase: Client = Depends(get_supabase)
):
    """Crea un nuovo scenario prompt"""
    try:
        # Verifica che il negozio appartenga al negoziante
        shop_result = supabase.table("shops").select("owner_id").eq("id", str(scenario.shop_id)).single().execute()
        if not shop_result.data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Negozio non trovato"
            )
        
        if shop_result.data["owner_id"] != current_user["id"]:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Puoi creare scenario prompts solo per i tuoi negozi"
            )
        
        scenario_data = {
            "shop_id": str(scenario.shop_id),
            "name": scenario.name,
            "description": scenario.description,
            "position": scenario.position,
            "environment": scenario.environment,
            "lighting": scenario.lighting,
            "background": scenario.background
        }
        
        result = supabase.table("scenario_prompts").insert(scenario_data).execute()
        
        if not result.data:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Errore durante la creazione dello scenario prompt"
            )
        
        return {
            "message": "Scenario prompt creato con successo",
            "scenario": result.data[0]
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Errore creazione scenario prompt: {e}")
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Errore durante la creazione dello scenario prompt: {str(e)}"
        )


@router.put("/{scenario_id}")
async def update_scenario_prompt(
    scenario_id: UUID,
    scenario: ScenarioPromptUpdate,
    current_user: dict = Depends(get_current_shop_owner),
    supabase: Client = Depends(get_supabase)
):
    """Aggiorna uno scenario prompt"""
    try:
        # Verifica permessi
        existing_result = supabase.table("scenario_prompts").select("shop_id").eq("id", str(scenario_id)).execute()
        if not existing_result.data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Scenario prompt non trovato"
            )
        
        shop_result = supabase.table("shops").select("owner_id").eq("id", existing_result.data[0]["shop_id"]).single().execute()
        if not shop_result.data or shop_result.data["owner_id"] != current_user["id"]:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Accesso negato"
            )
        
        updates = scenario.dict(exclude_unset=True)
        
        if not updates:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Nessun campo da aggiornare"
            )
        
        result = supabase.table("scenario_prompts").update(updates).eq("id", str(scenario_id)).execute()
        
        if not result.data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Scenario prompt non trovato"
            )
        
        return {
            "message": "Scenario prompt aggiornato con successo",
            "scenario": result.data[0]
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Errore aggiornamento scenario prompt: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Errore durante l'aggiornamento dello scenario prompt: {str(e)}"
        )


@router.delete("/{scenario_id}")
async def delete_scenario_prompt(
    scenario_id: UUID,
    current_user: dict = Depends(get_current_shop_owner),
    supabase: Client = Depends(get_supabase)
):
    """Elimina uno scenario prompt"""
    try:
        # Verifica permessi
        existing_result = supabase.table("scenario_prompts").select("shop_id").eq("id", str(scenario_id)).execute()
        if not existing_result.data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Scenario prompt non trovato"
            )
        
        shop_result = supabase.table("shops").select("owner_id").eq("id", existing_result.data[0]["shop_id"]).single().execute()
        if not shop_result.data or shop_result.data["owner_id"] != current_user["id"]:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Accesso negato"
            )
        
        result = supabase.table("scenario_prompts").delete().eq("id", str(scenario_id)).execute()
        
        return {
            "message": "Scenario prompt eliminato con successo",
            "scenario_id": str(scenario_id)
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Errore eliminazione scenario prompt: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Errore durante l'eliminazione dello scenario prompt: {str(e)}"
        )

File: frontend/pages/scenario_prompts.js
// Pagina gestione scenario prompts
(function() {
    'use strict';
    
    if (window.scenarioPromptsPageInitialized) {
        return;
    }
    window.scenarioPromptsPageInitialized = true;
    
    let currentScenarios = [];
    let currentShops = [];

    async function loadScenarioPrompts() {
        const container = document.getElementById('scenarios-list');
        if (!container) {
            console.warn('Container scenarios-list non trovato');
            return;
        }
        
        try {
            const data = await window.apiCall('/api/scenario-prompts/');
            currentScenarios = data.scenarios || [];
            renderScenarios();
        } catch (error) {
            console.error('Errore caricamento scenario prompts:', error);
            if (window.showError) {
                window.showError('Errore nel caricamento scenari: ' + error.message);
            }
        }
    }

    function renderScenarios() {
        const container = document.getElementById('scenarios-list');
        if (!container) return;
        
        if (currentScenarios.length === 0) {
            container.innerHTML = `
                <div class="empty-state">
                    <p>Nessuno scenario creato.</p>
                    <p>Crea il tuo primo scenario per iniziare!</p>
                </div>
            `;
            return;
        }
        
        container.innerHTML = currentScenarios.map(scenario => `
            <div class="scenario-card" style="border: 1px solid #ddd; border-radius: 8px; padding: 1.5rem; margin-bottom: 1rem; background: white;">
                <h3>${scenario.name}</h3>
                <p><strong>Descrizione:</strong> ${scenario.description}</p>
                ${scenario.position ? `<p><strong>Posizione:</strong> ${scenario.position}</p>` : ''}
                ${scenario.environment ? `<p><strong>Ambiente:</strong> ${scenario.environment}</p>` : ''}
                ${scenario.lighting ? `<p><strong>Illuminazione:</strong> ${scenario.lighting}</p>` : ''}
                ${scenario.background ? `<p><strong>Sfondo:</strong> ${scenario.background}</p>` : ''}
                <div style="margin-top: 1rem;">
                    <button onclick="editScenarioPrompt('${scenario.id}')" class="btn btn-small">Modifica</button>
                    <button onclick="deleteScenarioPrompt('${scenario.id}')" class="btn btn-small btn-danger">Elimina</button>
                </div>
            </div>
        `).join('');
    }

    async function loadShops() {
        try {
            const data = await window.apiCall('/api/shops/');
            currentShops = data.shops || data || [];
            return currentShops;
        } catch (error) {
            console.error('Errore caricamento negozi:', error);
            return [];
        }
    }

    function showCreateScenarioForm() {
        loadShops().then(shops => {
            if (shops.length === 0) {
                if (window.showError) {
                    window.showError('Devi prima creare un negozio!');
                }
                return;
            }
            
            const formHTML = `
                <div class="modal" id="scenario-modal">
                    <div class="modal-content" style="max-width: 700px;">
                        <span class="close" onclick="closeScenarioModal()">&times;</span>
                        <h2>Nuovo Scenario Prompt</h2>
                        <form id="scenario-form">
                            <div class="form-group">
                                <label>Negozio *</label>
                                <select id="scenario-shop" required>
                                    <option value="">Seleziona negozio...</option>
                                    ${shops.map(s => `<option value="${s.id}">${s.name}</option>`).join('')}
                                </select>
                            </div>
                            <div class="form-group">
                                <label>Nome Scenario *</label>
                                <input type="text" id="scenario-name" required placeholder="Es: Spiaggia Estate">
                            </div>
                            <div class="form-group">
                                <label>Descrizione *</label>
                                <textarea id="scenario-description" rows="3" required placeholder="Descrizione dettagliata della posizione, ambiente, illuminazione, ecc."></textarea>
                            </div>
                            <div class="form-group">
                                <label>Posizione</label>
                                <input type="text" id="scenario-position" placeholder="Es: in piedi, seduto, camminando">
                            </div>
                            <div class="form-group">
                                <label>Ambiente</label>
                                <input type="text" id="scenario-environment" placeholder="Es: interno, esterno, spiaggia, città">
                            </div>
                            <div class="form-group">
                                <label>Illuminazione</label>
                                <input type="text" id="scenario-lighting" placeholder="Es: naturale, artificiale, serale">
                            </div>
                            <div class="form-group">
                                <label>Sfondo</label>
                                <textarea id="scenario-background" rows="2" placeholder="Descrizione dello sfondo"></textarea>
                            </div>
                            <button type="submit" class="btn btn-primary">Crea Scenario</button>
                        </form>
                    </div>
                </div>
            `;
            
            document.body.insertAdjacentHTML('beforeend', formHTML);
            
            document.getElementById('scenario-form').addEventListener('submit', async (e) => {
                e.preventDefault();
                await createScenarioPrompt();
            });
        });
    }

    async function createScenarioPrompt() {
        const scenarioData = {
            shop_id: document.getElementById('scenario-shop').value,
            name: document.getElementById('scenario-name').value,
            description: document.getElementById('scenario-description').value,
            position: document.getElementById('scenario-position').value || null,
            environment: document.getElementById('scenario-environment').value || null,
            lighting: document.getElementById('scenario-lighting').value || null,
            background: document.getElementById('scenario-background').value || null
        };
        
        try {
            await window.apiCall('/api/scenario-prompts/', {
                method: 'POST',
                body: JSON.stringify(scenarioData)
            });
            
            if (window.showSuccess) {
                window.showSuccess('Scenario creato con successo!');
            }
            closeScenarioModal();
            loadScenarioPrompts();
        } catch (error) {
            if (window.showError) {
                window.showError('Errore nella creazione scenario: ' + error.message);
            }
        }
    }

    function editScenarioPrompt(scenarioId) {
        const scenario = currentScenarios.find(s => s.id === scenarioId);
        if (!scenario) {
            if (window.showError) {
                window.showError('Scenario non trovato');
            }
            return;
        }
        
        loadShops().then(shops => {
            const formHTML = `
                <div class="modal" id="scenario-modal">
                    <div class="modal-content" style="max-width: 700px;">
                        <span class="close" onclick="closeScenarioModal()">&times;</span>
                        <h2>Modifica Scenario Prompt</h2>
                        <form id="scenario-form">
                            <div class="form-group">
                                <label>Negozio *</label>
                                <select id="scenario-shop" required disabled>
                                    ${shops.map(s => `<option value="${s.id}" ${s.id === scenario.shop_id ? 'selected' : ''}>${s.name}</option>`).join('')}
                                </select>
                            </div>
                            <div class="form-group">
                                <label>Nome Scenario *</label>
                                <input type="text" id="scenario-name" required value="${scenario.name}">
                            </div>
                            <div class="form-group">
                                <label>Descrizione *</label>
                                <textarea id="scenario-description" rows="3" required>${scenario.description || ''}</textarea>
                            </div>
                            <div class="form-group">
                                <label>Posizione</label>
                                <input type="text" id="scenario-position" value="${scenario.position || ''}">
                            </div>
                            <div class="form-group">
                                <label>Ambiente</label>
                                <input type="text" id="scenario-environment" value="${scenario.environment || ''}">
                            </div>
                            <div class="form-group">
                                <label>Illuminazione</label>
                                <input type="text" id="scenario-lighting" value="${scenario.lighting || ''}">
                            </div>
                            <div class="form-group">
                                <label>Sfondo</label>
                                <textarea id="scenario-background" rows="2">${scenario.background || ''}</textarea>
                            </div>
                            <button type="submit" class="btn btn-primary">Salva Modifiche</button>
                        </form>
                    </div>
                </div>
            `;
            
            document.body.insertAdjacentHTML('beforeend', formHTML);
            
            document.getElementById('scenario-form').addEventListener('submit', async (e) => {
                e.preventDefault();
                await updateScenarioPrompt(scenarioId);
            });
        });
    }

    async function updateScenarioPrompt(scenarioId) {
        const scenarioData = {
            name: document.getElementById('scenario-name').value,
            description: document.getElementById('scenario-description').value,
            position: document.getElementById('scenario-position').value || null,
            environment: document.getElementById('scenario-environment').value || null,
            lighting: document.getElementById('scenario-lighting').value || null,
            background: document.getElementById('scenario-background').value || null
        };
        
        try {
            await window.apiCall(`/api/scenario-prompts/${scenarioId}`, {
                method: 'PUT',
                body: JSON.stringify(scenarioData)
            });
            
            if (window.showSuccess) {
                window.showSuccess('Scenario aggiornato con successo!');
            }
            closeScenarioModal();
            loadScenarioPrompts();
        } catch (error) {
            if (window.showError) {
                window.showError('Errore nell\'aggiornamento scenario: ' + error.message);
            }
        }
    }

    async function deleteScenarioPrompt(scenarioId) {
        if (!confirm('Sei sicuro di voler eliminare questo scenario?')) return;
        
        try {
            await window.apiCall(`/api/scenario-prompts/${scenarioId}`, { method: 'DELETE' });
            if (window.showSuccess) {
                window.showSuccess('Scenario eliminato con successo!');
            }
            loadScenarioPrompts();
        } catch (error) {
            if (window.showError) {
                window.showError('Errore nell\'eliminazione scenario: ' + error.message);
            }
        }
    }

    function closeScenarioModal() {
        const modal = document.getElementById('scenario-modal');
        if (modal) modal.remove();
    }

    // Esporta funzioni per uso globale
    window.loadScenarioPrompts = loadScenarioPrompts;
    window.showCreateScenarioForm = showCreateScenarioForm;
    window.editScenarioPrompt = editScenarioPrompt;
    window.deleteScenarioPrompt = deleteScenarioPrompt;
    window.closeScenarioModal = closeScenarioModal;

})();
